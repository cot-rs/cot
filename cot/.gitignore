# Generated files
static/admin/admin.css
templates/css/error.css
