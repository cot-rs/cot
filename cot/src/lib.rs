//! Cot is an easy to use, modern, and fast web framework for Rust. It has
//! been designed to be familiar if you've ever used
//! [Django](https://www.djangoproject.com/), and easy to learn if you haven't.
//! It's a batteries-included framework built on top of
//! [axum](https://github.com/tokio-rs/axum).
//!
//! ## Features
//!
//! * **Easy to use API** — in many ways modeled after Django, Cot's API is
//!   designed to be easy to use and intuitive. Sensible defaults make it for
//!   easy rapid development, while the API is still empowering you when needed.
//!   The documentation is a first-class citizen in Cot, making it easy to find
//!   what you're looking for.
//! * **ORM integration** — Cot comes with its own ORM, allowing you to interact
//!   with your database in a way that feels Rusty and intuitive. Rust types are
//!   the source of truth, and the ORM takes care of translating them to and
//!   from the database, as well as creating the migrations automatically.
//! * **Type safe** — wherever possible, Cot uses Rust's type system to prevent
//!   common mistakes and bugs. Not only views are taking advantage of the
//!   Rust's type system, but also the ORM, the admin panel, and even the
//!   templates. All that to catch errors as early as possible.
//! * **Admin panel** — Cot comes with an admin panel out of the box, allowing
//!   you to manage your app's data with ease. Adding new models to the admin
//!   panel is stupidly simple, making it a great tool not only for rapid
//!   development and debugging, but with its customization options, also for
//!   production use.
//! * **Secure by default** — security should be opt-out, not opt-in. Cot takes
//!   care of making your web apps secure by default, defending it against
//!   common modern web vulnerabilities. You can focus on building your app, not
//!   securing it.
//!
//! ## Guide
//!
//! This is an API reference for Cot, which might not be the best place to
//! start learning Cot. For a more gentle introduction, see the
//! [Cot guide](https://cot.rs/guide/latest/).
//!
//! ## Examples
//!
//! To see examples of how to use Cot, see the
//! [examples in the repository](https://github.com/cot-rs/cot/tree/master/examples).

#![warn(missing_docs, rustdoc::missing_crate_level_docs)]
#![cfg_attr(
    docsrs,
    feature(doc_cfg, doc_auto_cfg, rustdoc_missing_doc_code_examples)
)]
#![cfg_attr(docsrs, warn(rustdoc::missing_doc_code_examples))]
#![doc(
    html_logo_url = "https://raw.githubusercontent.com/cot-rs/media/6585c518/logo/logo.svg",
    html_favicon_url = "https://raw.githubusercontent.com/cot-rs/media/6585c518/logo/favicon.svg"
)]

extern crate self as cot;

#[cfg(feature = "db")]
pub mod db;
mod error;
pub mod form;
mod headers;
// Not public API. Referenced by macro-generated code.
#[doc(hidden)]
#[path = "private.rs"]
pub mod __private;
pub mod admin;
pub mod auth;
mod body;
pub mod cli;
pub mod config;
mod error_page;
mod handler;
pub mod html;
pub mod middleware;
#[cfg(feature = "openapi")]
pub mod openapi;
pub mod project;
pub mod request;
pub mod response;
pub mod router;
pub mod session;
pub mod static_files;
pub mod test;
pub(crate) mod utils;

#[cfg(feature = "openapi")]
pub use aide;
pub use body::Body;
pub use cot_macros::{main, test};
pub use error::Error;
#[cfg(feature = "openapi")]
pub use schemars;
pub use {bytes, http};

pub use crate::handler::{BoxedHandler, RequestHandler};
pub use crate::project::{
    App, AppBuilder, Bootstrapper, Project, ProjectContext, run, run_at, run_cli,
};

/// A type alias for a result that can return a [`cot::Error`].
pub type Result<T> = std::result::Result<T, Error>;

/// A type alias for an HTTP status code.
pub type StatusCode = http::StatusCode;

/// A type alias for an HTTP method.
pub type Method = http::Method;
