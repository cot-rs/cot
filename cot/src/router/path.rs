//! Path matching and routing.
//!
//! This module provides a path matcher that can be used to match paths against
//! a given pattern. It also provides a way to reverse paths to their original
//! form given a set of parameters.

use std::collections::HashMap;
use std::fmt::Display;

use thiserror::Error;
use tracing::debug;

#[derive(Debug, Clone)]
pub(super) struct PathMatcher {
    parts: Vec<PathPart>,
}

impl PathMatcher {
    #[must_use]
    pub(crate) fn new<T: Into<String>>(path_pattern: T) -> Self {
        #[derive(Debug, Copy, Clone)]
        enum State {
            Literal { start: usize },
            Param { start: usize },
        }

        let path_pattern = path_pattern.into();

        let mut parts = Vec::new();
        let mut state = State::Literal { start: 0 };

        let mut char_iter = path_pattern
            .chars()
            .map(Some)
            .chain([None])
            .enumerate()
            .peekable();
        loop {
            let Some((index, ch)) = char_iter.next() else {
                break;
            };

            match (ch, state) {
                (Some('{') | None, State::Literal { start }) => {
                    let literal = &path_pattern[start..index];
                    if literal.is_empty() {
                        assert!(
                            index == 0 || ch.is_none(),
                            "Consecutive parameters are not allowed"
                        );
                    } else {
                        parts.push(PathPart::Literal(literal.to_string()));
                    }
                    state = State::Param { start: index + 1 };
                }
                (Some('{'), State::Param { start }) => {
                    if start == index {
                        // escaped `{`
                        state = State::Literal { start: index };
                    } else {
                        panic!("Unclosed parameter: `{}`", &path_pattern[start..index]);
                    }
                }
                (Some('}'), State::Literal { start }) => {
                    let next_char = char_iter.peek().map(|(_, ch)| *ch).unwrap_or_default();

                    if next_char == Some('}') {
                        // escaped `}`
                        let literal = &path_pattern[start..=index];
                        parts.push(PathPart::Literal(literal.to_string()));

                        char_iter.next();
                        state = State::Literal { start: index + 2 };
                    } else {
                        panic!("Closing brace encountered without opening brace");
                    }
                }
                (Some('}'), State::Param { start }) => {
                    let param_name = &path_pattern[start..index].trim();
                    assert!(
                        Self::is_param_name_valid(param_name),
                        "Invalid parameter name: `{param_name}`"
                    );

                    parts.push(PathPart::Param {
                        name: (*param_name).to_string(),
                    });
                    state = State::Literal { start: index + 1 };
                }
                (Some('/') | None, State::Param { start }) => {
                    panic!("Unclosed parameter: `{}`", &path_pattern[start..index]);
                }
                _ => {}
            }
        }

        Self { parts }
    }

    fn is_param_name_valid(name: &str) -> bool {
        if name.is_empty() {
            return false;
        }
        let first_char = name.chars().next().expect("Empty string");
        if !first_char.is_alphabetic() && first_char != '_' {
            return false;
        }
        for ch in name.chars() {
            if !ch.is_alphanumeric() && ch != '_' {
                return false;
            }
        }
        true
    }

    #[must_use]
    pub(crate) fn capture<'matcher, 'path>(
        &'matcher self,
        path: &'path str,
    ) -> Option<CaptureResult<'matcher, 'path>> {
        debug!("Matching path `{}` against pattern `{}`", path, self);

        let mut current_path = path;
        let mut params = Vec::with_capacity(self.param_len());
        for part in &self.parts {
            match part {
                PathPart::Literal(s) => {
                    if !current_path.starts_with(s) {
                        return None;
                    }
                    current_path = &current_path[s.len()..];
                }
                PathPart::Param { name } => {
                    let next_slash = current_path.find('/');
                    let value = if let Some(next_slash) = next_slash {
                        &current_path[..next_slash]
                    } else {
                        current_path
                    };
                    if value.is_empty() {
                        return None;
                    }
                    params.push(PathParam::new(name, value));
                    current_path = &current_path[value.len()..];
                }
            }
        }

        Some(CaptureResult::new(params, current_path))
    }

    pub(crate) fn reverse(&self, params: &ReverseParamMap) -> Result<String, ReverseError> {
        let mut result = String::new();

        for part in &self.parts {
            match part {
                PathPart::Literal(s) => result.push_str(s),
                PathPart::Param { name } => {
                    let value = params
                        .get(name)
                        .ok_or_else(|| ReverseError::MissingParam(name.clone()))?;
                    result.push_str(value);
                }
            }
        }

        Ok(result)
    }

    #[must_use]
    fn param_len(&self) -> usize {
        self.param_names().count()
    }

    pub(super) fn param_names(&self) -> impl Iterator<Item = &str> {
        self.parts.iter().filter_map(|part| match part {
            PathPart::Literal(..) => None,
            PathPart::Param { name } => Some(name.as_str()),
        })
    }
}

impl Display for PathMatcher {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for part in &self.parts {
            write!(f, "{part}")?;
        }
        Ok(())
    }
}

/// A map of parameters for the [`crate::router::Router::reverse`] method.
///
/// Typically, it's only used internally via the [`crate::reverse`] macro.
///
/// # Examples
///
/// ```
/// use cot::router::path::ReverseParamMap;
///
/// let mut map = ReverseParamMap::new();
/// map.insert("id", "123");
/// map.insert("post_id", "456");
/// ```
#[derive(Debug)]
pub struct ReverseParamMap {
    params: HashMap<String, String>,
}

impl Default for ReverseParamMap {
    fn default() -> Self {
        Self::new()
    }
}

impl ReverseParamMap {
    /// Creates a new instance of [`ReverseParamMap`].
    ///
    /// # Examples
    ///
    /// ```
    /// use cot::router::path::ReverseParamMap;
    ///
    /// let mut map = ReverseParamMap::new();
    /// ```
    #[must_use]
    pub fn new() -> Self {
        Self {
            params: HashMap::new(),
        }
    }

    /// Inserts a value into the map. If the key already exists, the value will
    /// be overwritten.
    ///
    /// # Examples
    ///
    /// ```
    /// use cot::router::path::ReverseParamMap;
    ///
    /// let mut map = ReverseParamMap::new();
    /// map.insert("id", "123");
    /// map.insert("id", "456");
    /// ```
    #[expect(clippy::needless_pass_by_value)]
    pub fn insert<K: ToString, V: ToString>(&mut self, key: K, value: V) {
        self.params.insert(key.to_string(), value.to_string());
    }

    #[must_use]
    fn get(&self, key: &str) -> Option<&str> {
        self.params.get(key).map(String::as_str)
    }
}

#[doc(hidden)]
#[macro_export]
macro_rules! reverse_param_map {
    ($($key:ident = $value:expr),*) => {{
        #[allow(unused_mut)] // for the case when there are no parameters
        let mut map = $crate::router::path::ReverseParamMap::new();
        $( map.insert(stringify!($key), &$value); )*
        map
    }};
}

/// An error that occurs when reversing a path with missing parameters.
#[derive(Debug, Error)]
pub enum ReverseError {
    /// A parameter is missing for the reverse operation.
    #[error("Missing parameter for reverse: `{0}`")]
    MissingParam(String),
}

#[derive(Debug, PartialEq, Eq)]
pub(super) struct CaptureResult<'matcher, 'path> {
    pub(super) params: Vec<PathParam<'matcher>>,
    pub(super) remaining_path: &'path str,
}

impl<'matcher, 'path> CaptureResult<'matcher, 'path> {
    #[must_use]
    fn new(params: Vec<PathParam<'matcher>>, remaining_path: &'path str) -> Self {
        Self {
            params,
            remaining_path,
        }
    }

    #[must_use]
    pub(crate) fn matches_fully(&self) -> bool {
        self.remaining_path.is_empty()
    }
}

#[derive(Debug, Clone)]
enum PathPart {
    Literal(String),
    Param { name: String },
}

impl Display for PathPart {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PathPart::Literal(s) => {
                let s = s.replace('{', "{{").replace('}', "}}");
                write!(f, "{s}")
            }
            PathPart::Param { name } => write!(f, "{{{name}}}"),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(super) struct PathParam<'a> {
    pub(super) name: &'a str,
    pub(super) value: String,
}

impl<'a> PathParam<'a> {
    #[must_use]
    pub(crate) fn new(name: &'a str, value: &str) -> Self {
        Self {
            name,
            value: value.to_string(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reverse_param_map_default() {
        let map = ReverseParamMap::default();
        assert_eq!(map.params.len(), 0);
    }

    #[test]
    fn path_parser_no_params() {
        let path_parser = PathMatcher::new("/users");
        assert_eq!(
            path_parser.capture("/users"),
            Some(CaptureResult::new(vec![], ""))
        );
        assert_eq!(path_parser.capture("/test"), None);
    }

    #[test]
    fn path_parser_escaped() {
        let path_parser = PathMatcher::new("/users/{{{{{{escaped}}}}}}");
        assert_eq!(
            path_parser.capture("/users/{{{escaped}}}"),
            Some(CaptureResult::new(vec![], ""))
        );
    }

    #[test]
    fn path_parser_single_param() {
        let path_parser = PathMatcher::new("/users/{id}");
        assert_eq!(
            path_parser.capture("/users/123"),
            Some(CaptureResult::new(vec![PathParam::new("id", "123")], ""))
        );
        assert_eq!(
            path_parser.capture("/users/123/"),
            Some(CaptureResult::new(vec![PathParam::new("id", "123")], "/"))
        );
        assert_eq!(
            path_parser.capture("/users/123/abc"),
            Some(CaptureResult::new(
                vec![PathParam::new("id", "123")],
                "/abc"
            ))
        );
        assert_eq!(path_parser.capture("/users/"), None);
    }

    #[test]
    fn path_parser_param_whitespace() {
        let path_parser = PathMatcher::new("/users/{ id }");

        assert_eq!(
            path_parser.capture("/users/123"),
            Some(CaptureResult::new(vec![PathParam::new("id", "123")], ""))
        );
    }

    #[test]
    fn path_parser_multiple_params() {
        let path_parser = PathMatcher::new("/users/{id}/posts/{post_id}");
        assert_eq!(
            path_parser.capture("/users/123/posts/456"),
            Some(CaptureResult::new(
                vec![
                    PathParam::new("id", "123"),
                    PathParam::new("post_id", "456"),
                ],
                ""
            ))
        );
        assert_eq!(
            path_parser.capture("/users/123/posts/456/abc"),
            Some(CaptureResult::new(
                vec![
                    PathParam::new("id", "123"),
                    PathParam::new("post_id", "456"),
                ],
                "/abc"
            ))
        );
    }

    #[test]
    #[should_panic(expected = "Consecutive parameters are not allowed")]
    fn path_parser_consecutive_params() {
        let _ = PathMatcher::new("/users/{id}{post_id}");
    }

    #[test]
    #[should_panic(expected = "Invalid parameter name: ``")]
    fn path_parser_invalid_name_empty() {
        let _ = PathMatcher::new("/users/{}");
    }

    #[test]
    #[should_panic(expected = "Invalid parameter name: `123`")]
    fn path_parser_invalid_name_numeric() {
        let _ = PathMatcher::new("/users/{123}");
    }

    #[test]
    #[should_panic(expected = "Invalid parameter name: `abc#$%`")]
    fn path_parser_invalid_name_non_alphanumeric() {
        let _ = PathMatcher::new("/users/{abc#$%}");
    }

    #[test]
    #[should_panic(expected = "Unclosed parameter: `foo`")]
    fn path_parser_unclosed() {
        let _ = PathMatcher::new("/users/{foo");
    }

    #[test]
    #[should_panic(expected = "Closing brace encountered without opening brace")]
    fn path_parser_missing_opening_brace() {
        let _ = PathMatcher::new("/users/foo}");
    }

    #[test]
    #[should_panic(expected = "Unclosed parameter: `foo`")]
    fn path_parser_unclosed_slash() {
        let _ = PathMatcher::new("/users/{foo/bar");
    }

    #[test]
    #[should_panic(expected = "Unclosed parameter: `foo`")]
    fn path_parser_unclosed_double() {
        let _ = PathMatcher::new("/users/{foo{bar");
    }

    #[test]
    #[should_panic(expected = "Closing brace encountered without opening brace")]
    fn path_parser_escaping_unclosed() {
        let _ = PathMatcher::new("/users/{{{foo}}/bar");
    }

    #[test]
    fn path_parser_display() {
        let path_parser = PathMatcher::new("/users/{id}/posts/{{escaped}}");
        assert_eq!(format!("{path_parser}"), "/users/{id}/posts/{{escaped}}");
    }

    #[test]
    fn reverse_with_valid_params() {
        let path_parser = PathMatcher::new("/users/{id}/posts/{post_id}");
        let mut params = ReverseParamMap::new();
        params.insert("id", "123");
        params.insert("post_id", "456");
        assert_eq!(
            path_parser.reverse(&params).unwrap(),
            "/users/123/posts/456"
        );
    }

    #[test]
    fn reverse_with_missing_param() {
        let path_parser = PathMatcher::new("/users/{id}/posts/{post_id}");
        let mut params = ReverseParamMap::new();
        params.insert("id", "123");
        let result = path_parser.reverse(&params);
        assert!(result.is_err());
        assert_eq!(
            result.unwrap_err().to_string(),
            "Missing parameter for reverse: `post_id`"
        );
    }

    #[test]
    fn reverse_with_extra_param() {
        let path_parser = PathMatcher::new("/users/{id}/posts/{post_id}");
        let mut params = ReverseParamMap::new();
        params.insert("id", "123");
        params.insert("post_id", "456");
        params.insert("extra", "789");
        assert_eq!(
            path_parser.reverse(&params).unwrap(),
            "/users/123/posts/456"
        );
    }
}
