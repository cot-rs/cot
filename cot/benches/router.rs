use criterion::{Criterion, criterion_group, criterion_main};
use serde::{Deserialize, Serialize};

mod bench_utils;
use bench_utils::bench;
use cot::json::Json;
use cot::router::{Route, Router};

async fn hello_world() -> &'static str {
    "Hello, World!"
}

// JSON API endpoint
#[derive(Debug, Serialize, Deserialize)]
struct AddRequest {
    a: i32,
    b: i32,
}

#[derive(Debug, Serialize, Deserialize)]
struct AddResponse {
    result: i32,
}

async fn add_numbers(Json(req): Json<AddRequest>) -> Json<AddResponse> {
    Json(AddResponse {
        result: req.a + req.b,
    })
}

pub fn criterion_benchmark(c: &mut Criterion) {
    bench(c, "empty_router")
        .expected_status_code(reqwest::StatusCode::NOT_FOUND)
        .run_with_router(Router::empty);

    bench(c, "single_root_route")
        .path("/")
        .run_with_router(|| Router::with_urls([Route::with_handler("/", hello_world)]));

    bench(c, "single_root_route_burst")
        .path("/")
        .requests_per_iteration(1000)
        .run_with_router(|| Router::with_urls([Route::with_handler("/", hello_world)]));

    bench(c, "nested_routers")
        .path("/a/b/c/d/e/f/g")
        .run_with_router(|| {
            Router::with_urls([Route::with_router(
                "/a",
                Router::with_urls([Route::with_router(
                    "/b",
                    Router::with_urls([Route::with_router(
                        "/c",
                        Router::with_urls([Route::with_router(
                            "/d",
                            Router::with_urls([Route::with_router(
                                "/e",
                                Router::with_urls([Route::with_router(
                                    "/f",
                                    Router::with_urls([Route::with_handler("/g", hello_world)]),
                                )]),
                            )]),
                        )]),
                    )]),
                )]),
            )])
        });

    bench(c, "json_api")
        .path("/")
        .method(reqwest::Method::POST)
        .json_body(&AddRequest { a: 10, b: 20 })
        .run_with_router(|| Router::with_urls([Route::with_handler("/", add_numbers)]));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
