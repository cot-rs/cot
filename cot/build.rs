use std::path::PathBuf;

#[cfg(all(
    feature = "db",
    not(any(feature = "sqlite", feature = "postgres", feature = "mysql"))
))]
compile_error!("feature \"db\" requires one of: \"sqlite\", \"postgres\", \"mysql\" to be enabled");

fn main() {
    build_css();
}

fn build_css() {
    const SCSS_FILES: [(&str, &str); 2] = [
        ("admin/admin.scss", "static/admin/admin.css"),
        ("error.scss", "templates/css/error.css"),
    ];

    let options = scss_options();

    for (scss_file, css_file) in SCSS_FILES {
        let scss_path = format!("scss/{scss_file}");

        println!("cargo::rerun-if-changed={scss_path}");

        let css = grass::from_path(scss_path, &options).expect("failed to compile SCSS");

        let out_dir = std::env::var("OUT_DIR").expect("OUT_DIR should be set");
        let css_path = PathBuf::from(out_dir).join(css_file);
        let css_dir = css_path
            .parent()
            .expect("failed to get CSS parent directory");
        std::fs::create_dir_all(css_dir).expect("failed to create CSS directory");
        std::fs::write(css_path, css).expect("failed to write CSS");
    }
}

fn scss_options() -> grass::Options<'static> {
    grass::Options::default().style(grass::OutputStyle::Compressed)
}
