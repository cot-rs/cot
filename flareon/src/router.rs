use std::collections::HashMap;
use std::fmt::{Debug, Formatter};
use std::sync::Arc;

use axum::http::StatusCode;
use bytes::Bytes;
use log::debug;

use crate::request::Request;
use crate::router::path::{PathMatcher, ReverseParamMap};
use crate::{Body, Error, RequestHandler, Response, Result};

pub mod path;

#[derive(Clone, Debug)]
pub struct Router {
    urls: Vec<Route>,
    names: HashMap<String, Arc<PathMatcher>>,
}

impl Router {
    #[must_use]
    pub fn with_urls<T: Into<Vec<Route>>>(urls: T) -> Self {
        let urls = urls.into();
        let mut names = HashMap::new();

        for url in &urls {
            if let Some(name) = &url.name {
                names.insert(name.clone(), url.url.clone());
            }
        }

        Self { urls, names }
    }

    async fn route(&self, mut request: Request, request_path: &str) -> Result<Response> {
        debug!("Routing request to {}", request_path);

        for route in &self.urls {
            if let Some(matches) = route.url.capture(request_path) {
                let matches_fully = matches.matches_fully();
                for param in matches.params {
                    request
                        .path_params
                        .insert(param.name.to_owned(), param.value);
                }

                match &route.view {
                    RouteInner::Handler(handler) => {
                        if matches_fully {
                            return handler.handle(request).await;
                        }
                    }
                    RouteInner::Router(router) => {
                        return Box::pin(router.route(request, matches.remaining_path)).await
                    }
                }
            }
        }

        debug!("Not found: {}", request_path);
        Ok(handle_not_found())
    }

    /// Handle a request.
    ///
    /// This method is called by the [`FlareonApp`](crate::FlareonApp) to handle
    /// a request.
    ///
    /// # Errors
    ///
    /// This method re-throws any errors that occur in the request handler.
    pub async fn handle(&self, request: Request) -> Result<Response> {
        let path = request.uri().path().to_owned();
        self.route(request, &path).await
    }

    /// Get a URL for a view by name.
    ///
    /// Instead of using this method directly, consider using the
    /// [`reverse!`](crate::reverse) macro which provides much more
    /// ergonomic way to call this.
    ///
    /// # Errors
    ///
    /// This method returns an error if the view name is not found.
    ///
    /// This method returns an error if the URL cannot be generated because of
    /// missing parameters.
    pub fn reverse(&self, name: &str, params: &ReverseParamMap) -> Result<String> {
        self.reverse_option(name, params)?
            .ok_or_else(|| Error::NoViewToReverse {
                view_name: name.to_owned(),
            })
    }

    /// Get a URL for a view by name.
    ///
    /// Returns `None` if the view name is not found.
    ///
    /// # Errors
    ///
    /// This method returns an error if the URL cannot be generated because of
    /// missing parameters.
    pub fn reverse_option(&self, name: &str, params: &ReverseParamMap) -> Result<Option<String>> {
        let url = self.names.get(name).map(|matcher| matcher.reverse(params));
        if let Some(url) = url {
            return Ok(Some(url?));
        }

        for route in &self.urls {
            if let RouteInner::Router(router) = &route.view {
                if let Some(url) = router.reverse_option(name, params)? {
                    return Ok(Some(route.url.reverse(params)? + &url));
                }
            }
        }
        Ok(None)
    }

    #[must_use]
    pub fn routes(&self) -> &[Route] {
        &self.urls
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.urls.is_empty()
    }
}

#[derive(Debug, Clone)]
pub struct Route {
    url: Arc<PathMatcher>,
    view: RouteInner,
    name: Option<String>,
}

impl Route {
    #[must_use]
    pub fn with_handler<V: RequestHandler + Send + Sync + 'static>(url: &str, view: V) -> Self {
        Self {
            url: Arc::new(PathMatcher::new(url)),
            view: RouteInner::Handler(Arc::new(Box::new(view))),
            name: None,
        }
    }

    #[must_use]
    pub fn with_handler_and_name<T: Into<String>, V: RequestHandler + Send + Sync + 'static>(
        url: &str,
        view: V,
        name: T,
    ) -> Self {
        Self {
            url: Arc::new(PathMatcher::new(url)),
            view: RouteInner::Handler(Arc::new(Box::new(view))),
            name: Some(name.into()),
        }
    }

    #[must_use]
    pub fn with_router(url: &str, router: Router) -> Self {
        Self {
            url: Arc::new(PathMatcher::new(url)),
            view: RouteInner::Router(router),
            name: None,
        }
    }
}

fn handle_not_found() -> Response {
    Response::new_html(
        StatusCode::NOT_FOUND,
        Body::Fixed(Bytes::from("404 Not Found")),
    )
}

#[derive(Clone)]
enum RouteInner {
    Handler(Arc<Box<dyn RequestHandler + Send + Sync>>),
    Router(Router),
}

impl Debug for RouteInner {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match &self {
            RouteInner::Handler(_) => f.debug_tuple("Handler").field(&"handler(...)").finish(),
            RouteInner::Router(router) => f.debug_tuple("Router").field(router).finish(),
        }
    }
}

/// Reverse a URL for a view by name and given params.
#[macro_export]
macro_rules! reverse_str {
    ($request:expr, $view_name:literal $(, $($key:expr => $value:expr),*)?) => {
        $request
            .project()
            .router()
            .reverse($view_name, &$crate::reverse_param_map!($( $($key => $value),* )?))?
    };
}

/// Reverse a URL for a view by name and given params and return a response with
/// a redirect.
///
/// This macro is a shorthand for creating a response with a redirect to a URL
/// generated by the [`reverse_str!`] macro.
#[macro_export]
macro_rules! reverse {
    ($request:expr, $view_name:literal $(, $($key:expr => $value:expr),*)?) => {
        $crate::Response::new_redirect($crate::reverse_str!(
            $request,
            $view_name,
            $( $($key => $value),* )?
        ))
    };
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use flareon::FlareonProject;

    use super::*;
    use crate::request::Request;
    use crate::Response;

    struct MockHandler;

    #[async_trait::async_trait]
    impl RequestHandler for MockHandler {
        async fn handle(&self, _request: Request) -> Result<Response> {
            Ok(Response::new_html(
                StatusCode::OK,
                Body::Fixed(Bytes::from("OK")),
            ))
        }
    }

    #[test]
    fn test_router_with_urls() {
        let route = Route::with_handler("/test", MockHandler);
        let router = Router::with_urls(vec![route.clone()]);
        assert_eq!(router.routes().len(), 1);
    }

    #[tokio::test]
    async fn test_router_route() {
        let route = Route::with_handler("/test", MockHandler);
        let router = Router::with_urls(vec![route.clone()]);
        let response = router.route(test_request(), "/test").await.unwrap();
        assert_eq!(response.status, StatusCode::OK);
    }

    #[tokio::test]
    async fn test_router_handle() {
        let route = Route::with_handler("/test", MockHandler);
        let router = Router::with_urls(vec![route.clone()]);
        let response = router.handle(test_request()).await.unwrap();
        assert_eq!(response.status, StatusCode::OK);
    }

    #[test]
    fn test_router_reverse() {
        let route = Route::with_handler_and_name("/test", MockHandler, "test");
        let router = Router::with_urls(vec![route.clone()]);
        let params = ReverseParamMap::new();
        let url = router.reverse("test", &params).unwrap();
        assert_eq!(url, "/test");
    }

    #[test]
    fn test_router_reverse_with_param() {
        let route = Route::with_handler_and_name("/test/:id", MockHandler, "test");
        let router = Router::with_urls(vec![route.clone()]);
        let mut params = ReverseParamMap::new();
        params.insert("id", "123");
        let url = router.reverse("test", &params).unwrap();
        assert_eq!(url, "/test/123");
    }

    #[test]
    fn test_router_reverse_option() {
        let route = Route::with_handler_and_name("/test", MockHandler, "test");
        let router = Router::with_urls(vec![route.clone()]);
        let params = ReverseParamMap::new();
        let url = router.reverse_option("test", &params).unwrap().unwrap();
        assert_eq!(url, "/test");
    }

    #[test]
    fn test_router_routes() {
        let route = Route::with_handler("/test", MockHandler);
        let router = Router::with_urls(vec![route.clone()]);
        assert_eq!(router.routes().len(), 1);
    }

    #[test]
    fn test_router_is_empty() {
        let router = Router::with_urls(vec![]);
        assert!(router.is_empty());
    }

    #[test]
    fn test_route_with_handler() {
        let route = Route::with_handler("/test", MockHandler);
        assert_eq!(route.url.to_string(), "/test");
    }

    #[test]
    fn test_route_with_handler_and_params() {
        let route = Route::with_handler("/test/:id", MockHandler);
        assert_eq!(route.url.to_string(), "/test/:id");
    }

    #[test]
    fn test_route_with_handler_and_name() {
        let route = Route::with_handler_and_name("/test", MockHandler, "test");
        assert_eq!(route.url.to_string(), "/test");
        assert_eq!(route.name.as_deref(), Some("test"));
    }

    #[test]
    fn test_route_with_router() {
        let sub_route = Route::with_handler("/sub", MockHandler);
        let sub_router = Router::with_urls(vec![sub_route]);
        let route = Route::with_router("/test", sub_router);
        assert_eq!(route.url.to_string(), "/test");
    }

    fn test_request() -> Request {
        let request = Request::new(
            axum::http::Request::builder()
                .uri("/test")
                .body(axum::body::Body::empty())
                .unwrap(),
            Arc::new(FlareonProject::builder().build()),
        );
        request
    }
}
