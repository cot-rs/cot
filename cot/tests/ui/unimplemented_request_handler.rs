use cot::router::{Route, Router};

async fn test(_invalid: ()) -> cot::Result<cot::response::Response> {
    unimplemented!()
}

fn main() {
    let _ = Router::with_urls([Route::with_handler("/", test)]);
}
