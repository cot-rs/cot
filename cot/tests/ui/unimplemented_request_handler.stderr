error[E0277]: `fn(()) -> impl Future<Output = Result<cot::http::Response<Body>, cot::Error>> {test}` is not a valid request handler
 --> tests/ui/unimplemented_request_handler.rs:8:57
  |
8 |     let _ = Router::with_urls([Route::with_handler("/", test)]);
  |                                -------------------      ^^^^ not a valid request handler
  |                                |
  |                                required by a bound introduced by this call
  |
  = help: the trait `RequestHandler<_>` is not implemented for fn item `fn(()) -> impl Future<Output = Result<cot::http::Response<Body>, cot::Error>> {test}`
  = note: make sure the function is marked `async`
  = note: make sure all parameters implement `FromRequest` or `FromRequestParts`
  = note: make sure there is at most one parameter implementing `FromRequest`
  = note: make sure the function takes no more than 10 parameters
  = help: the following other types implement trait `RequestHandler<T>`:
            `ApiMethodRouter` implements `RequestHandler`
            `MethodRouter` implements `RequestHandler`
            `cot::openapi::NoApi<H>` implements `RequestHandler<HandlerParams>`
note: required by a bound in `Route::with_handler`
 --> src/router.rs
  |
  |     pub fn with_handler<HandlerParams, H>(url: &str, handler: H) -> Self
  |            ------------ required by a bound in this associated function
...
  |         H: RequestHandler<HandlerParams> + Send + Sync + 'static,
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Route::with_handler`
